update infrastructure.rst/compilers (#1946)
conda-forge/conda-forge.github.io: src/maintainer/infrastructure.rst

frequency and scope, so it is not a current concern. However, we keep our policies
for such ABI breaks in place for the next time it should occur.

While we do not have any promises of support for a generation of ABI-compatible
compilers, we have historically maintained them according to the following (non-binding)
principles.

* The authoritative source of the current compilers and versions for various languages
  and platforms is the `conda_build_config.yaml <https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/conda_build_config.yaml>`_
  in the `conda-forge/conda-forge-pinning-feedstock <https://github.com/conda-forge/conda-forge-pinning-feedstock>`_
  as described in :ref:`globally_pinned_packages`.
* We provide no support of any kind in terms of the long-term stability / support of a given compiler generation.
* We upgrade them in an ad-hoc manner on a periodic basis as we have the time and energy to do so.
  Note that because of the way we enforce runtime constraints, these compiler upgrades will not break
  existing packages. However, if you are using the compilers outside of ``conda``, then you may find issues.

> * We will provide notice in the form of an announcement when an ABI-incompatible compiler change is going to be happen.

Reviewer: ```suggestion
* We usually provide notice in the form of an announcement when an ABI-incompatible compiler change is going to happen.
```

> * We usually provide notice in the form of an announcement when an ABI-incompatible compiler change is going to happen.

Reviewer 2: Can you please keep this as it was? It was agreed upon in a core meeting and I don't want to change it.

Reviewer: Yes will do. My apologies. Not right this minute though.

Reviewer: Done. https://github.com/conda-forge/conda-forge.github.io/pull/1947

Author: > Done. https://github.com/conda-forge/conda-forge.github.io/pull/1947

We're not in such a rush IMO... I didn't press merge here, I'd be happy to discuss this some more. Aside from a typo ("conpiler"), #1947 is IMO not correct, because that section now explicitly tries to draw a distinction between ABI-breaking and ABI-preserving compiler upgrades, and you made that distinction less clear.

Reviewer: I'll fix the typo ofc but I'm not gonna split hairs on the language here. Isuru is correct that this language was agreed upon in a core meeting and so should remain unchanged. I mentioned this in the previous pr as well at some point iirc. 

Please accept my apologies for merging too early. The pr looked good to go and so I went ahead. 

I'm happy to back out the prs or we can continue debate in a new one. 

I really appreciate the hard work from you @h-vetinari on this. 

One thing I realized is that we don't separate out what is an informal policy from core and what is more general documentation. That is making this process more difficult than it has to be.

Author: > this language was agreed upon in a core meeting and so should remain unchanged

I disagree with the tenet that something that has been (core-)agreed once is forever unchangeable, especially if circumstances change.

You can back out my PR (or not 🤷); my interest here is not to offend but to provide useful documentation, in particular the distinction between ABI-breaking and not ABI-breaking compiler upgrades[^1]. If that needs a core vote, then that's fine by me, but I wrote this text specifically with that separation in mind. If people don't want that separation, just revert.

[^1]: as an instance of the latter is what started this when John [asked](https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/4215#issuecomment-1518358829) for a docs update

that form the given stack.

The more recent status quo is that compilers can be upgraded in an ABI-compatible manner,
meaning that we can just increase the version in our global pinning, and it will slowly
roll out in a compatible manner to the ecosystem as feedstocks get rerendered.
For such ABI-compatible upgrades, similar but looser principles apply:

* The pins are similarly in the global pinning, see :ref:`globally_pinned_packages`.
* We provide no support of any kind in terms of the long-term availability of a given compiler generation.
* We generally provide notice in the form of an announcement when a compiler is going to be upgraded.
* In general, our compilers on Linux and OSX are using very recent compilers, whereas
  on windows, we generally use the last supported VS version.

Despite the lack of explicit support, we try to keep the compilers in their various versions
working also outside of conda-forge, and even provide an easy way to install them
(through the `compilers <https://github.com/conda-forge/compilers-feedstock>`_ feedstock).

In more detail, our default compiler stack is made up very differently on each platform.

> More specifically, each compiler needs an _activation_ that makes the difference

Reviewer: ```suggestion
More specifically, each compiler uses an _activation_ package that makes the difference
```

OSX (Clang):

* [C, C++] Activation: https://github.com/conda-forge/clang-compiler-activation-feedstock/
* [C, C++] Required feedstocks:
  `llvmdev <https://github.com/conda-forge/llvmdev-feedstock>`_,
  `clangdev <https://github.com/conda-forge/clangdev-feedstock>`_,
  `compiler-rt <https://github.com/conda-forge/compiler-rt-feedstock>`_,
  `libcxx <https://github.com/conda-forge/libcxx-feedstock>`_,
  `openmp <https://github.com/conda-forge/openmp-feedstock>`_,
  `lld <https://github.com/conda-forge/lld-feedstock>`_,
  `cctools <https://github.com/conda-forge/cctools-and-ld64-feedstock>`_
* [Fortran] Activation: https://github.com/conda-forge/gfortran_osx-64-feedstock/
* [Fortran] Implementation: https://github.com/conda-forge/gfortran_impl_osx-64-feedstock/

Windows (MSVC):

* [C, C++] Activation: https://github.com/conda-forge/vc-feedstock
  (we cannot redistribute the actual MSVC compilers due to licensing constraints)
* [Fortran] Activation & Implementation: https://github.com/conda-forge/flang-feedstock

There exists an alternative, MinGW-based, compiler stack on windows, which is available

> with a ``m2w64_`` prefix (e.g. ``{{ compiler('m2w64_c') }}``), which is however falling out

Reviewer: ```suggestion
with a ``m2w64_`` prefix (e.g. ``{{ compiler('m2w64_c') }}``). However, it is falling out
```

of use now that most projects will natively support compilation also with MSVC, in addition
to several complications arising from mixing compiler stacks.

> Additionally, there is a possibility to use clang as a compiler also on Linux & Windows:

Reviewer: ```suggestion
Additionally, there is a possibility to use clang as a compiler on Linux & Windows:
```

  and platforms is the `conda_build_config.yaml <https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/conda_build_config.yaml>`_
  in the `conda-forge/conda-forge-pinning-feedstock <https://github.com/conda-forge/conda-forge-pinning-feedstock>`_
  as described in :ref:`globally_pinned_packages`.
* We provide no support of any kind in terms of the long-term stability of these pinnings.
* We upgrade them in an ad-hoc manner on a periodic basis as we have the time and energy to do so.
  Note that because of the way we enforce runtime constraints, these compiler upgrades will not break
  existing packages. However, if you are using the compilers outside of ``conda``, then you may find issues.
* We generally provide notice in the form of an announcement when a compiler is going to be upgraded.
  Note that these changes take a bit of time to complete, so you will generally have time
  to prepare should you need to.
* Some of the criteria we think about when considering a compiler migration include
  1) the degree of disruption to the ecosystem, 2) the amount of work for the ``core`` team,

>   and 3) the amount of time it will cost our (volunteer) feedstock maintainers.

Reviewer: A lot of the language here was agreed upon by the core team and should stay.

Author: Sure. I've put them back in with the exception of the last two, which aren't relevant anymore, since we don't have to migrate to change the compiler version.

Author: Though I'd argue that even the middle two points are not relevant anymore either (and the first one I had kept in the running text).

* We upgrade them in an ad-hoc manner on a periodic basis as we have the time and energy to do so.
  Note that because of the way we enforce runtime constraints, these compiler upgrades will not break
  existing packages. However, if you are using the compilers outside of ``conda``, then you may find issues.
* We generally provide notice in the form of an announcement when a compiler is going to be upgraded.
  Note that these changes take a bit of time to complete, so you will generally have time

>   to prepare should you need to.

Reviewer 2: Why?

Author: Are you asking why I removed this?

Because AFAIK the last gcc & clang updates didn't come with an announcement, and that in turn because it's no as big of a deal anymore as changing to comp7 (which is noted further below in that document as "our current compiler stack") was once upon a time.

Nowadays, gcc & clang bumps just percolate through the ecosystem at the speed with which feedstocks get rerendered, without compatibility constraints (well, aside from small issues here and there, but not large-scale ABI incompatibilties).

Reviewer 2: This is not about ABI compatibilites and if you look at the git history it is not about gcc 4->7 migration.

> aside from small issues here and there, but not large-scale ABI incompatibilties

Exactly why announcements are there

Author: > Exactly why announcements are there

If someone had asked me for an announcement for https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/4215, I would have done it, no problem. However, neither were [clang 13](https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/2802), [clang 14](https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/3261) or [GCC 11](https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/3731) announced; also it's the first time during all those bumps (from what I saw) that someone notices a breakage at all.

> This is not about ABI compatibilites and if you look at the git history it is not about gcc 4->7 migration.

OK, I'll readd the announcement bit. However the "Note that these changes take a bit of time to complete, so you will generally have time to prepare should you need to." is not true anymore AFAICT.
